// select-pure/lib/components/Option.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, LitElement, html } from "lit";
import { customElement } from "lit/decorators/custom-element.js";
import { property } from "lit/decorators/property.js";
import { ifDefined } from "lit-html/directives/if-defined.js";
import { KEYS } from "./constants";
let OptionPure = class OptionPure extends LitElement {
    constructor() {
        super();
        this._selected = false;
        this._disabled = false;
        this._hidden = false;
        this._value = "";
        this._label = "";
        this.optionIndex = -1;
        // bindings
        this.onClick = this.onClick.bind(this);
        this.select = this.select.bind(this);
        this.unselect = this.unselect.bind(this);
        this.getOption = this.getOption.bind(this);
    }
    static get styles() {
        return css `
      .option {
        align-items: center;
        background-color: var(--background-color, #fff);
        box-sizing: border-box;
        color: var(--color, #000);
        cursor: pointer;
        display: flex;
        font-family: var(--font-family, inherit);
        font-size: var(--font-size, 14px);
        font-weight: var(--font-weight, 400);
        height: var(--select-height, 44px);
        height: var(--select-height, 44px);
        justify-content: flex-start;
        padding: var(--padding, 0 10px);
        width: 100%;
      }
      .option:not(.disabled):focus, .option:not(.disabled):not(.selected):hover {
        background-color: var(--hover-background-color, #e3e3e3);
        color: var(--hover-color, #000);
      }
      .selected {
        background-color: var(--selected-background-color, #e3e3e3);
        color: var(--selected-color, #000);
      }
      .disabled {
        background-color: var(--disabled-background-color, #e3e3e3);
        color: var(--disabled-color, #000);
        cursor: default;
      }
    `;
    }
    connectedCallback() {
        super.connectedCallback();
        // set properties
        this._selected = this.getAttribute("selected") !== null;
        this._disabled = this.getAttribute("disabled") !== null;
        this._hidden = this.getAttribute("hidden") !== null;
        this._value = this.getAttribute("value") || "";
        this.processLabel();
        if (this.onReady) {
            this.onReady(this.getOption(), this.optionIndex);
        }
    }
    getOption() {
        return {
            label: this._label,
            value: this._value,
            select: this.select,
            unselect: this.unselect,
            selected: this._selected,
            disabled: this._disabled,
            hidden: this._hidden,
        };
    }
    select() {
        this._selected = true;
        this.setAttribute("selected", "");
    }
    unselect() {
        this._selected = false;
        this.removeAttribute("selected");
    }
    setOnReadyCallback(callback, index) {
        this.onReady = callback;
        this.optionIndex = index;
    }
    setOnSelectCallback(callback) {
        this.onSelect = callback;
    }
    processLabel() {
        if (this.textContent) {
            this._label = this.textContent;
            return;
        }
        if (this.getAttribute("label")) {
            this._label = this.getAttribute("label") || "";
        }
    }
    onClick(event) {
        event.stopPropagation();
        if (!this.onSelect || this._disabled) {
            return;
        }
        this.onSelect(this._value);
    }
    handleKeyPress(event) {
        if (event.which === KEYS.ENTER) {
            this.onClick(event);
        }
    }
    render() {
        const classNames = ["option"];
        if (this._selected) {
            classNames.push("selected");
        }
        if (this._disabled) {
            classNames.push("disabled");
        }
        return html `
      <div
        class="${classNames.join(" ")}"
        @click=${this.onClick}
        @keydown="${this.handleKeyPress}"
        tabindex="${ifDefined(this._disabled ? undefined : "0")}"
      >
        <slot hidden @slotchange=${this.processLabel}></slot>
        ${this._label}
      </div>
    `;
    }
};
__decorate([
    property()
], OptionPure.prototype, "_selected", void 0);
__decorate([
    property()
], OptionPure.prototype, "_disabled", void 0);
__decorate([
    property()
], OptionPure.prototype, "_hidden", void 0);
__decorate([
    property()
], OptionPure.prototype, "_value", void 0);
__decorate([
    property()
], OptionPure.prototype, "_label", void 0);
__decorate([
    property()
], OptionPure.prototype, "optionIndex", void 0);
OptionPure = __decorate([
    customElement("option-pure")
], OptionPure);
export { OptionPure };
